robot/web/stream/server.go

```diff
@@ -319,6 +329,104 @@ func (server *Server) RemoveStream(ctx context.Context, req *streampb.RemoveStre
+// properties.
+func (server *Server) GetStreamOptions(
+	ctx context.Context,
+	req *streampb.GetStreamOptionsRequest,
+) (*streampb.GetStreamOptionsResponse, error) {
+	if req.Name == "" {
+		return nil, errors.New("stream name is required")
+	}
+	cam, err := camera.FromRobot(server.robot, req.Name)
+	if err != nil {
+		return nil, err
+	}
+	// If the camera properties do not have intrinsic parameters,
+	// we need to sample a frame to get the width and height.
+	var width, height int
+	camProps, err := cam.Properties(ctx)
+	if err != nil || camProps.IntrinsicParams == nil || camProps.IntrinsicParams.Width == 0 || camProps.IntrinsicParams.Height == 0 {
+		server.logger.Debug("width and height not found in camera properties")
+		width, height, err = server.sampleFrameSize(ctx, cam)
+		if err != nil {
+			return nil, fmt.Errorf("failed to sample frame size: %w", err)
+		}
+	} else {
+		width, height = camProps.IntrinsicParams.Width, camProps.IntrinsicParams.Height
+	}
+	scaledResolutions := server.generateResolutions(int32(width), int32(height))
+	resolutions := make([]*streampb.Resolution, 0, 5)
+	for _, res := range scaledResolutions {
+		resolutions = append(resolutions, &streampb.Resolution{
+			Height: res.height,
+			Width:  res.width,
+		})
+	}
+	return &streampb.GetStreamOptionsResponse{
+		Resolutions: resolutions,
+	}, nil
+}
+
+// generateResolutions takes the original width and height of an image and returns
+// a list of the original resolution with 4 smaller width/height options that maintain
+// the same aspect ratio.
+func (server *Server) generateResolutions(width, height int32) []resolution {
+	resolutions := []resolution{
+		{width: width, height: height},
+	}
+	// We use integer division to get the scaled width and height. Fractions are truncated
+	// to the nearest integer. This means that the scaled width and height may not match the
+	// original aspect ratio exactly if source dimensions are odd.
+	for i := 0; i < 4; i++ {
+		width /= 2
+		height /= 2
+		resolutions = append(resolutions, resolution{width: width, height: height})
+		server.logger.Debugf("scaled resolution %d: %dx%d", i, width, height)
+	}
+	return resolutions
+}
+
+// sampleFrameSize takes in a camera.Camera, starts a stream, attempts to
+// pull a frame using Stream.Next, and returns the width and height.
+func (server *Server) sampleFrameSize(ctx context.Context, cam camera.Camera) (int, int, error) {
+	server.logger.Debug("sampling frame size")
+	stream, err := cam.Stream(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer func() {
+		if cerr := stream.Close(ctx); cerr != nil {
+			server.logger.Error("failed to close stream:", cerr)
+		}
+	}()
+	// Attempt to get a frame from the stream with a maximum of 5 retries.
+	// This is useful if cameras have a warm-up period before they can start streaming.
+	var frame image.Image
+	var release func()
+	for i := 0; i < 5; i++ {
```

Comment:
```suggestion
	for i := range 5 {
```

Reply:
Unless rdk's linters isn't bumped up for go, then leave as is.

Reply:
Not [1.23](https://eli.thegreenplace.net/2024/ranging-over-functions-in-go-123/), though I think this is fine if the linter doesn't complain at you, which I assumed it was since we bumped to 1.23, but I guess we disabled that sometime in the recent past. Which is good, this is more explicit anyway.